File: data/src/main/java/com/imadji/arch/data/source/remote/interceptor/RequestInterceptor.java
package com.imadji.arch.data.source.remote.interceptor;

import java.io.IOException;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class RequestInterceptor implements Interceptor {
    private String apiKey;

    public RequestInterceptor(String apiKey) {
        this.apiKey = apiKey;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request originalRequest = chain.request();
        HttpUrl originalHttpUrl = originalRequest.url();

        HttpUrl newUrl = originalHttpUrl.newBuilder()
                .addQueryParameter("api_key", apiKey)
                .build();

        Request.Builder requestBuilder = originalRequest.newBuilder()
                .url(newUrl);

        Request newRequest = requestBuilder.build();

        return chain.proceed(newRequest);
    }
}

File: data/src/main/java/com/imadji/arch/data/source/remote/interceptor/RequestInterceptor.kt
package com.imadji.arch.data.source.remote.interceptor

import okhttp3.Interceptor
import okhttp3.Response
import javax.inject.Inject

class RequestInterceptor @Inject constructor(private val apiKey: String) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val originalRequest = chain.request()
        val originalHttpUrl = originalRequest.url

        val newUrl = originalHttpUrl.newBuilder()
                .addQueryParameter(REQUEST_PARAM_API_KEY, apiKey)
                .build()

        val requestBuilder = originalRequest.newBuilder()
                .url(newUrl)

        val newRequest = requestBuilder.build()

        return chain.proceed(newRequest)
    }

    companion object {
        private const val REQUEST_PARAM_API_KEY = "api_key"
    }
}